File: pygento/agento_image_server.py
#!/usr/bin/env python3
"""
Flask server for image generation using Gemini API.

This server provides the same functionality as agento_image.py CLI tool,
but accepts JSON payloads via HTTP endpoints.
"""
import os
import json
import traceback
from flask import Flask, request, jsonify, send_file
from pathlib import Path
import tempfile
import shutil

# Import the service classes from agento_image.py
from agento_image import NanaBabanaImageService, generate_descriptive_filename

app = Flask(__name__)

# Configuration
DEFAULT_PORT = int(os.getenv('PORT', 5000))
DEFAULT_HOST = os.getenv('HOST', '0.0.0.0')
GEMINI_API_KEY = os.getenv('GEMINI_API_KEY', '')
BASE_PATH = Path(os.getenv('BASE_PATH', Path.cwd()))
UPLOAD_FOLDER = BASE_PATH / 'pub' / 'media' / 'lookbook'
UPLOAD_FOLDER.mkdir(parents=True, exist_ok=True)

# Global service instance
image_service = NanaBabanaImageService(
    api_key=GEMINI_API_KEY,
    base_path=str(BASE_PATH)
)

@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint."""
    return jsonify({
        'status': 'healthy',
        'service': 'agento-image-server',
        'api_available': image_service.is_available()
    })

@app.route('/generate', methods=['POST'])
def generate_image():
    """
    Generate image from one or two input images.

    Expected JSON payload:
    {
        "model_image": "path/to/image.jpg or https://url/to/image.jpg",
        "look_image": "path/to/image.jpg or https://url/to/image.jpg (optional)",
        "prompt": "Description of what to generate",
        "api_key": "optional override API key"
    }
    """
    try:
        # Get JSON data
        data = request.get_json()
        if not data:
            return jsonify({
                'success': False,
                'error': 'No JSON payload provided'
            }), 400

        # Extract parameters
        model_image = data.get('model_image')
        look_image = data.get('look_image')  # Optional
        prompt = data.get('prompt')
        api_key = data.get('api_key', GEMINI_API_KEY)

        # Validate required parameters
        if not model_image:
            return jsonify({
                'success': False,
                'error': 'model_image is required'
            }), 400

        if not prompt:
            return jsonify({
                'success': False,
                'error': 'prompt is required'
            }), 400

        # Create service instance with optional API key override
        service = NanaBabanaImageService(
            api_key=api_key,
            base_path=str(BASE_PATH)
        )

        if not service.is_available():
            return jsonify({
                'success': False,
                'error': 'API key not configured'
            }), 500

        # Generate image
        op = service.generate_image(model_image, look_image, prompt)

        if op.get('done'):
            # Generate descriptive filename
            cache_key = generate_descriptive_filename(model_image, look_image, prompt)

            # Save the generated image
            saved_path = service.save_asset_from_operation(op, cache_key=cache_key)

            return jsonify({
                'success': True,
                'status': 'completed',
                'saved_path': saved_path,
                'filename': Path(saved_path).name,
                'message': f'Image generated and saved to {saved_path}',
                'input': {
                    'model_image': model_image,
                    'look_image': look_image,
                    'prompt': prompt
                }
            })
        else:
            # Async response (fallback)
            return jsonify(op)

    except Exception as e:
        app.logger.error(f"Error generating image: {str(e)}")
        app.logger.error(traceback.format_exc())
        return jsonify({
            'success': False,
            'error': str(e),
            'traceback': traceback.format_exc()
        }), 500

@app.route('/images/<filename>', methods=['GET'])
def get_image(filename):
    """Serve generated images."""
    try:
        image_path = UPLOAD_FOLDER / filename
        if not image_path.exists():
            return jsonify({
                'success': False,
                'error': f'Image {filename} not found'
            }), 404

        return send_file(str(image_path), mimetype='image/jpeg')

    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@app.errorhandler(404)
def not_found(error):
    return jsonify({
        'success': False,
        'error': 'Endpoint not found'
    }), 404

@app.errorhandler(500)
def internal_error(error):
    return jsonify({
        'success': False,
        'error': 'Internal server error'
    }), 500

def main():
    """Run the Flask server."""
    print(f"Starting Agento Image Server on {DEFAULT_HOST}:{DEFAULT_PORT}")
    print(f"Upload folder: {UPLOAD_FOLDER}")
    print(f"API available: {image_service.is_available()}")

    app.run(
        host=DEFAULT_HOST,
        port=DEFAULT_PORT,
        debug=os.getenv('DEBUG', 'false').lower() == 'true'
    )

if __name__ == '__main__':
    main()

File: nodegento/agento_image_server.js
#!/usr/bin/env node
/**
 * Agento Image Server - Node.js
 * HTTP server for image generation using Gemini API
 */

const express = require('express');
const path = require('path');
const fs = require('fs');
const { NanaBabanaImageService } = require('./agento_image_service');
const { generateDescriptiveFilename } = require('./agento_image_filename');

const app = express();
const PORT = process.env.PORT || 3000;
const HOST = process.env.HOST || '0.0.0.0';

// Middleware
app.use(express.json({ limit: '50mb' }));
app.use(express.urlencoded({ extended: true, limit: '50mb' }));

// Configuration
const GEMINI_API_KEY = process.env.GEMINI_API_KEY || '';
const BASE_PATH = process.env.BASE_PATH || process.cwd();
const UPLOAD_FOLDER = path.join(BASE_PATH, 'pub', 'media', 'lookbook');

// Global service instance
const imageService = new NanaBabanaImageService(
    GEMINI_API_KEY,
    BASE_PATH,
    null,
    process.env.DEBUG === 'true'
);

// Ensure upload directory exists
if (!fs.existsSync(UPLOAD_FOLDER)) {
    fs.mkdirSync(UPLOAD_FOLDER, { recursive: true });
}

// Routes

app.get('/health', (req, res) => {
    /** Health check endpoint. */
    res.json({
        status: 'healthy',
        service: 'agento-image-server-node',
        api_available: imageService.isAvailable()
    });
});

app.post('/generate', async (req, res) => {
    /**
     * Generate image from one or two input images.
     *
     * Expected JSON payload:
     * {
     *   "model_image": "path/to/image.jpg or https://url/to/image.jpg",
     *   "look_image": "path/to/image.jpg or https://url/to/image.jpg (optional)",
     *   "prompt": "Description of what to generate",
     *   "api_key": "optional override API key"
     * }
     */
    try {
        const { model_image, look_image, prompt, api_key } = req.body;

        // Validate required parameters
        if (!model_image) {
            return res.status(400).json({
                success: false,
                error: 'model_image is required'
            });
        }

        if (!prompt) {
            return res.status(400).json({
                success: false,
                error: 'prompt is required'
            });
        }

        // Create service instance with optional API key override
        const service = new NanaBabanaImageService(
            api_key || GEMINI_API_KEY,
            BASE_PATH
        );

        if (!service.isAvailable()) {
            return res.status(500).json({
                success: false,
                error: 'API key not configured'
            });
        }

        // Generate image
        const op = await service.generateImage(model_image, look_image, prompt);

        if (op.done) {
            // Generate descriptive filename
            const cacheKey = generateDescriptiveFilename(model_image, look_image, prompt);

            // Save the generated image
            const savedPath = service.saveAssetFromOperation(op, cacheKey);

            res.json({
                success: true,
                status: 'completed',
                saved_path: savedPath,
                filename: path.basename(savedPath),
                message: `Image generated and saved to ${savedPath}`,
                input: {
                    model_image,
                    look_image,
                    prompt
                }
            });
        } else {
            // Async response (fallback)
            res.json(op);
        }

    } catch (error) {
        console.error('Error generating image:', error);
        res.status(500).json({
            success: false,
            error: error.message,
            stack: process.env.DEBUG === 'true' ? error.stack : undefined
        });
    }
});

app.get('/images/:filename', (req, res) => {
    /** Serve generated images. */
    try {
        const filename = req.params.filename;
        const imagePath = path.join(UPLOAD_FOLDER, filename);

        if (!fs.existsSync(imagePath)) {
            return res.status(404).json({
                success: false,
                error: `Image ${filename} not found`
            });
        }

        res.sendFile(imagePath);

    } catch (error) {
        res.status(500).json({
            success: false,
            error: error.message
        });
    }
});

// Error handling
app.use((err, req, res, next) => {
    console.error(err.stack);
    res.status(500).json({
        success: false,
        error: 'Internal server error'
    });
});

// 404 handler
app.use((req, res) => {
    res.status(404).json({
        success: false,
        error: 'Endpoint not found'
    });
});

// Start server
if (require.main === module) {
    app.listen(PORT, HOST, () => {
        console.log(`🚀 Agento Image Server (Node.js) listening on ${HOST}:${PORT}`);
        console.log(`📁 Upload folder: ${UPLOAD_FOLDER}`);
        console.log(`🔑 API available: ${imageService.isAvailable()}`);
        if (process.env.DEBUG === 'true') {
            console.log('🐛 Debug mode enabled');
        }
    });
}

module.exports = app;